docstr_coverage/result_collection.py

```diff
@@ -0,0 +1,312 @@
+        try:
+            return self._files[file_path]
+        except KeyError:
+            file = File()
+            self._files[file_path] = file
+            return file
+
+    def count(self):
+        """Walks through all the tracked files in this result collection, and counts overall
+        statistics, such as #missing docstring.
+
+        Returns
+        -------
+        AggregatedCount
+            A count instance containing a range of docstring counts."""
+        counts = (folder.count() for folder in self._files.values())
+        return functools.reduce(operator.add, counts, AggregatedCount())
+
+    def files(self):
+        """Generator, iterating over all (file-name, file-info) tuples in this result collection"""
+        for file_path, file in self._files.items():
+            yield file_path, file
+
+    def to_legacy(self):
+        """Converts the information in this `ResultCollection` into the less expressive dictionary
+        of counts used since the early versions of docstr-coverage."""
+        file_results = dict()
+        for filename, file in self.files():
+            missing_list = [
+                e.node_identifier
+                for e in file.expected_docstrings()
+                if not e.ignore_reason
+                and not e.has_docstring
+                and not e.node_identifier == "module docstring"
+            ]
+            has_module_doc = (
+                len(
+                    [
+                        e
+                        for e in file.expected_docstrings()
+                        if e.node_identifier == "module docstring" and e.has_docstring
+                    ]
+                )
+                > 0
+            )
+            count = file.count()
+            file_results[filename] = {
+                "missing": missing_list,
+                "module_doc": has_module_doc,
+                "missing_count": count.missing,
+                "needed_count": count.needed,
+                "coverage": count.coverage(),
+                "empty": count.is_empty,
+            }
+        total_count = self.count()
+        total_results = {
+            "missing_count": total_count.missing,
+            "needed_count": total_count.needed,
+            "coverage": total_count.coverage(),
+        }
+        return file_results, total_results
+
+
+class FileStatus(enum.Enum):
+    """Upon inspection by docstr-coverage,
+    every file is assigned exactly one of the states from this enum."""
+
+    ANALYZED = (1,)
+    EMPTY = 2
+    # Note: We may eventually add IGNORED here, but this will require some more refactoring
+
+
+class File:
+    """The information about docstrings for a single file."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._expected_docstrings = []
+        self._status = FileStatus.ANALYZED
+
+    def report(self, identifier: str, has_docstring: bool, ignore_reason: str = None):
```

Comment:
I think `collect`/`record_docstring` might be a more appropriate name for this method than `report`. I usually thing of "reporting" as being a lot like logging, and this method just collects items to be reported on later. If this change makes sense to you, the `report_module` method below would need to be changed as well